# Import

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("merged_comments.csv")

In [3]:
df

,video_id,comment_id,parent_comment_id,is_reply,comment_text,comment_published_at,comment_updated_at,like_count,reply_count,author_channel_id,collected_at
0,R4-zhQcOecM,UgysxUh1THZ9nQqJaMB4AaABAg,NaN,False,Cho mình biết bạn muốn mình đến quán bẫy du lị...,2026-09-06T11:03:29Z,2026-09-06T11:03:29Z,357,136,UC04uCYXt-E-gsBBuEG3kLew,2026-09-18T08:12:52.586097+00:00
1,R4-zhQcOecM,UgxXDluAQvvhdC40ga54AaABAg,NaN,False,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",2026-09-17T14:27:43Z,2026-09-17T14:27:43Z,0,0,UCjPyKfzZw_pRm0OTo83HJPQ,2026-09-18T08:12:52.586097+00:00
2,R4-zhQcOecM,Ugy4Hpy8u9vQIsRU8eN4AaABAg,NaN,False,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,2026-09-17T13:02:49Z,2026-09-17T13:02:49Z,0,0,UCHhc-QgZAAVrzXL_ILpmYmQ,2026-09-18T08:12:52.586097+00:00
3,R4-zhQcOecM,UgzveMhZ_NaaGxXCiFR4AaABAg,NaN,False,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",2026-09-17T11:45:19Z,2026-09-17T11:45:19Z,0,0,UCa7-Fz3J4jsnMdQ-P1L14RA,2026-09-18T08:12:52.586097+00:00
4,R4-zhQcOecM,UgyMy7oNdvD2Opwlh8d4AaABAg,NaN,False,nice,2026-09-17T08:27:14Z,2026-09-17T08:27:14Z,0,0,UC7KFZSWH82DKaNH2uCT5j8w,2026-09-18T08:12:52.586097+00:00
...,...,...,...,...,...,...,...,...,...,...,...
24343,lQRIvQ28wVA,UgxjH9-MKlPO1ZyR9Y94AaABAg,NaN,False,Nice place!😍,2021-12-25 14:11:01+00:00,2021-12-25T14:11:01Z,1,0,UC5aczqXyfmQk2k0GBxb9ayg,2026-09-18T09:17:49.905232+00:00
24344,lQRIvQ28wVA,Ugz6CMvXjElVk_3Fks94AaABAg,NaN,False,"super uplaod, Joel. )",2021-12-23 09:48:48+00:00,2021-12-23T09:48:48Z,2,1,UCsrDZMfP9xvY374d8RMf2Kw,2026-09-18T09:17:49.905257+00:00
24345,lQRIvQ28wVA,UgyBSWRWA8n8MdmgG1N4AaABAg,NaN,False,A wonderful MALAYSIA!,2021-12-23 09:19:56+00:00,2021-12-23T09:19:56Z,2,1,UCWeXFP1x6lsGeKGhDHgkM5w,2026-09-18T09:17:49.905267+00:00
24346,lQRIvQ28wVA,UgyoYTpR0e4Bi3K0jqF4AaABAg,NaN,False,Nice vlog dude.,2021-12-23 09:18:36+00:00,2021-12-23T09:18:36Z,2,1,UCoEQik-7LIBrvKSs0Ju9aIw,2026-09-18T09:17:49.905276+00:00


record = df['comment_published_at']
usable = []
temp=0
for row in record :
  if row < "2026-01-01T00:00:00Z":
    temp+=1
    usable.append(row)

print(temp)
usableC = pd.DataFrame(usable)
usableC.sort_values(by=0, ascending=True)

df['comment_published_at'] = pd.to_datetime(df['comment_published_at'], errors='coerce')

df = df[
    (df['comment_published_at'] >= '2015-01-01') &
    (df['comment_published_at'] < '2026-01-01')
].copy()

df.reset_index(drop=True, inplace=True)

print("Total records exclude 2026:", len(df))

df

In [4]:
print(df.isna().sum().sort_values(ascending=False))

parent_comment_id       19106
comment_text                9
video_id                    0
comment_id                  0
is_reply                    0
comment_published_at        0
comment_updated_at          0
like_count                  0
reply_count                 0
author_channel_id           0
collected_at                0
dtype: int64


In [5]:
import re
import pandas as pd

def clean_comment(text):

    if pd.isna(text):
        return ""

    text = str(text)

    # Remove URLs
    text = re.sub(
        r"http\S+|www\S+",
        " ",
        text
    )

    # Remove excessive whitespace
    text = re.sub(
        r"\s+",
        " ",
        text
    )

    return text.strip()


df["clean_text"] = df["comment_text"].apply(
    clean_comment
)

In [6]:
df

,video_id,comment_id,parent_comment_id,is_reply,comment_text,comment_published_at,comment_updated_at,like_count,reply_count,author_channel_id,collected_at,clean_text
0,R4-zhQcOecM,UgysxUh1THZ9nQqJaMB4AaABAg,NaN,False,Cho mình biết bạn muốn mình đến quán bẫy du lị...,2026-09-06T11:03:29Z,2026-09-06T11:03:29Z,357,136,UC04uCYXt-E-gsBBuEG3kLew,2026-09-18T08:12:52.586097+00:00,Cho mình biết bạn muốn mình đến quán bẫy du lị...
1,R4-zhQcOecM,UgxXDluAQvvhdC40ga54AaABAg,NaN,False,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",2026-09-17T14:27:43Z,2026-09-17T14:27:43Z,0,0,UCjPyKfzZw_pRm0OTo83HJPQ,2026-09-18T08:12:52.586097+00:00,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top"
2,R4-zhQcOecM,Ugy4Hpy8u9vQIsRU8eN4AaABAg,NaN,False,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,2026-09-17T13:02:49Z,2026-09-17T13:02:49Z,0,0,UCHhc-QgZAAVrzXL_ILpmYmQ,2026-09-18T08:12:52.586097+00:00,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...
3,R4-zhQcOecM,UgzveMhZ_NaaGxXCiFR4AaABAg,NaN,False,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",2026-09-17T11:45:19Z,2026-09-17T11:45:19Z,0,0,UCa7-Fz3J4jsnMdQ-P1L14RA,2026-09-18T08:12:52.586097+00:00,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ..."
4,R4-zhQcOecM,UgyMy7oNdvD2Opwlh8d4AaABAg,NaN,False,nice,2026-09-17T08:27:14Z,2026-09-17T08:27:14Z,0,0,UC7KFZSWH82DKaNH2uCT5j8w,2026-09-18T08:12:52.586097+00:00,nice
...,...,...,...,...,...,...,...,...,...,...,...,...
24343,lQRIvQ28wVA,UgxjH9-MKlPO1ZyR9Y94AaABAg,NaN,False,Nice place!😍,2021-12-25 14:11:01+00:00,2021-12-25T14:11:01Z,1,0,UC5aczqXyfmQk2k0GBxb9ayg,2026-09-18T09:17:49.905232+00:00,Nice place!😍
24344,lQRIvQ28wVA,Ugz6CMvXjElVk_3Fks94AaABAg,NaN,False,"super uplaod, Joel. )",2021-12-23 09:48:48+00:00,2021-12-23T09:48:48Z,2,1,UCsrDZMfP9xvY374d8RMf2Kw,2026-09-18T09:17:49.905257+00:00,"super uplaod, Joel. )"
24345,lQRIvQ28wVA,UgyBSWRWA8n8MdmgG1N4AaABAg,NaN,False,A wonderful MALAYSIA!,2021-12-23 09:19:56+00:00,2021-12-23T09:19:56Z,2,1,UCWeXFP1x6lsGeKGhDHgkM5w,2026-09-18T09:17:49.905267+00:00,A wonderful MALAYSIA!
24346,lQRIvQ28wVA,UgyoYTpR0e4Bi3K0jqF4AaABAg,NaN,False,Nice vlog dude.,2021-12-23 09:18:36+00:00,2021-12-23T09:18:36Z,2,1,UCoEQik-7LIBrvKSs0Ju9aIw,2026-09-18T09:17:49.905276+00:00,Nice vlog dude.


In [7]:
df = df[
    df["clean_text"].str.len() > 2
].copy()

In [8]:
df = df.drop_duplicates(
    subset=["comment_id"]
)

In [9]:
df = df.drop_duplicates(
    subset=["clean_text"]
)

In [10]:
print("Remaining comments:", len(df))

Remaining comments: 23389


In [11]:
df["word_count"] = (
    df["clean_text"]
    .str.findall(r"\b[\w']+\b")
    .str.len()
)
df["word_count"].describe()

,word_count
count,23389.000000
mean,19.183206
std,29.587839
min,0.000000
25%,5.000000
50%,11.000000
75%,22.000000
max,1156.000000


In [12]:
df

,video_id,comment_id,parent_comment_id,is_reply,comment_text,comment_published_at,comment_updated_at,like_count,reply_count,author_channel_id,collected_at,clean_text,word_count
0,R4-zhQcOecM,UgysxUh1THZ9nQqJaMB4AaABAg,NaN,False,Cho mình biết bạn muốn mình đến quán bẫy du lị...,2026-09-06T11:03:29Z,2026-09-06T11:03:29Z,357,136,UC04uCYXt-E-gsBBuEG3kLew,2026-09-18T08:12:52.586097+00:00,Cho mình biết bạn muốn mình đến quán bẫy du lị...,17
1,R4-zhQcOecM,UgxXDluAQvvhdC40ga54AaABAg,NaN,False,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",2026-09-17T14:27:43Z,2026-09-17T14:27:43Z,0,0,UCjPyKfzZw_pRm0OTo83HJPQ,2026-09-18T08:12:52.586097+00:00,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",13
2,R4-zhQcOecM,Ugy4Hpy8u9vQIsRU8eN4AaABAg,NaN,False,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,2026-09-17T13:02:49Z,2026-09-17T13:02:49Z,0,0,UCHhc-QgZAAVrzXL_ILpmYmQ,2026-09-18T08:12:52.586097+00:00,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,14
3,R4-zhQcOecM,UgzveMhZ_NaaGxXCiFR4AaABAg,NaN,False,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",2026-09-17T11:45:19Z,2026-09-17T11:45:19Z,0,0,UCa7-Fz3J4jsnMdQ-P1L14RA,2026-09-18T08:12:52.586097+00:00,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",18
4,R4-zhQcOecM,UgyMy7oNdvD2Opwlh8d4AaABAg,NaN,False,nice,2026-09-17T08:27:14Z,2026-09-17T08:27:14Z,0,0,UC7KFZSWH82DKaNH2uCT5j8w,2026-09-18T08:12:52.586097+00:00,nice,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...
24343,lQRIvQ28wVA,UgxjH9-MKlPO1ZyR9Y94AaABAg,NaN,False,Nice place!😍,2021-12-25 14:11:01+00:00,2021-12-25T14:11:01Z,1,0,UC5aczqXyfmQk2k0GBxb9ayg,2026-09-18T09:17:49.905232+00:00,Nice place!😍,2
24344,lQRIvQ28wVA,Ugz6CMvXjElVk_3Fks94AaABAg,NaN,False,"super uplaod, Joel. )",2021-12-23 09:48:48+00:00,2021-12-23T09:48:48Z,2,1,UCsrDZMfP9xvY374d8RMf2Kw,2026-09-18T09:17:49.905257+00:00,"super uplaod, Joel. )",3
24345,lQRIvQ28wVA,UgyBSWRWA8n8MdmgG1N4AaABAg,NaN,False,A wonderful MALAYSIA!,2021-12-23 09:19:56+00:00,2021-12-23T09:19:56Z,2,1,UCWeXFP1x6lsGeKGhDHgkM5w,2026-09-18T09:17:49.905267+00:00,A wonderful MALAYSIA!,3
24346,lQRIvQ28wVA,UgyoYTpR0e4Bi3K0jqF4AaABAg,NaN,False,Nice vlog dude.,2021-12-23 09:18:36+00:00,2021-12-23T09:18:36Z,2,1,UCoEQik-7LIBrvKSs0Ju9aIw,2026-09-18T09:17:49.905276+00:00,Nice vlog dude.,3


In [13]:
df[
    df["word_count"] <= 2
][
    ["clean_text", "word_count"]
]

,clean_text,word_count
4,nice,1
5,Hay quá,2
7,❤❤❤❤❤,0
26,🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣,0
32,GHFFHDTGFFFFFFFFGG,1
...,...,...
24310,myvi kinnggu,2
24316,Best youtuber,2
24340,Nice one! 👍,2
24343,Nice place!😍,2


In [14]:
df[df["word_count"]<=2]["clean_text"].to_csv("text<2.csv", index=False)

In [15]:
# make sure clean_text is string
df["clean_text"] = df["clean_text"].fillna("").astype(str).str.strip()

def remove_mentions(text):
    return re.sub(r"@\S+", "", text).strip()


def has_text(text):
    """
    Check whether comment contains at least
    one letter or number.
    """
    return any(char.isalnum() for char in text)


def classify_basic(text):

    text_no_mention = remove_mentions(text)
    text_lower = text_no_mention.lower().strip()

    # -----------------------------------------
    # 1. emoji / punctuation / mention only
    # -----------------------------------------

    if not has_text(text_no_mention):
        return "irrelevant_noise"


    # -----------------------------------------
    # 2. obvious conversational noise
    # -----------------------------------------

    noise_words = {
        "thanks",
        "thank you",
        "thank u",
        "tq",
        "tqtq",
        "first",
        "im first",
        "i'm first",
        "hello",
        "hi",
        "please",
        "welcome",
        "ok",
        "okay",
        "agreed",
        "sure",
        "haha",
        "hahaha"
    }

    # Remove punctuation/emojis for comparison
    simple_text = re.sub(
        r"[^a-zA-Z0-9\s']",
        "",
        text_lower
    ).strip()

    if simple_text in noise_words:
        return "irrelevant_noise"


    # -----------------------------------------
    # 3. tourism related keywords
    # -----------------------------------------

    tourism_keywords = [

        # Malaysia / destinations
        "malaysia",
        "malaysian",
        "kuala lumpur",
        "penang",
        "langkawi",
        "melaka",
        "malacca",
        "johor",
        "sabah",
        "sarawak",
        "borneo",
        "putrajaya",
        "sipadan",
        "tioman",
        "perhentian",
        "genting",
        "cameron highlands",
        "kinabalu",
        "semporna",
        "kuching",

        # Tourism
        "travel",
        "travelling",
        "traveling",
        "tour",
        "tourist",
        "tourism",
        "trip",
        "visit",
        "visited",
        "vlog",

        # Nature / attractions
        "beach",
        "island",
        "mountain",
        "park",
        "view",
        "views",
        "scenery",
        "waterfall",
        "nature",
        "diving",
        "snorkeling",
        "sea",
        "coast",
        "wildlife",

        # Accommodation / transport
        "hotel",
        "hostel",
        "resort",
        "airport",
        "flight",
        "train",
        "bus",
        "grab",
        "taxi",

        # Cost
        "price",
        "cheap",
        "expensive",
        "ringgit",

        # Food
        "food",
        "restaurant",
        "nasi lemak",
        "delicious",
        "spicy",

        # Malay / Indonesian terms
        "tempat",
        "cantik",
        "cantek",
        "menarik",
        "sedap",
        "makanan",
        "mahal",
        "murah",
        "pantai",
        "pulau",
        "pelancong",
        "wisata",
        "indah"
    ]

    for keyword in tourism_keywords:

        if keyword in text_lower:
            return "tourism_relevant"


    # -----------------------------------------
    # 4. generic sentiment
    # -----------------------------------------

    sentiment_words = [
        "good",
        "great",
        "beautiful",
        "amazing",
        "awesome",
        "bad",
        "terrible",
        "love",
        "hate",
        "best",
        "worst",
        "perfect",
        "underrated",
        "overrated"
    ]

    for word in sentiment_words:

        if re.search(r"\b" + re.escape(word) + r"\b", text_lower):
            return "sentiment_only"


    # -----------------------------------------
    # 5. can't decide
    # -----------------------------------------

    return "uncertain"


df["relevance"] = df["clean_text"].apply(classify_basic)

df

,video_id,comment_id,parent_comment_id,is_reply,comment_text,comment_published_at,comment_updated_at,like_count,reply_count,author_channel_id,collected_at,clean_text,word_count,relevance
0,R4-zhQcOecM,UgysxUh1THZ9nQqJaMB4AaABAg,NaN,False,Cho mình biết bạn muốn mình đến quán bẫy du lị...,2026-09-06T11:03:29Z,2026-09-06T11:03:29Z,357,136,UC04uCYXt-E-gsBBuEG3kLew,2026-09-18T08:12:52.586097+00:00,Cho mình biết bạn muốn mình đến quán bẫy du lị...,17,uncertain
1,R4-zhQcOecM,UgxXDluAQvvhdC40ga54AaABAg,NaN,False,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",2026-09-17T14:27:43Z,2026-09-17T14:27:43Z,0,0,UCjPyKfzZw_pRm0OTo83HJPQ,2026-09-18T08:12:52.586097+00:00,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",13,uncertain
2,R4-zhQcOecM,Ugy4Hpy8u9vQIsRU8eN4AaABAg,NaN,False,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,2026-09-17T13:02:49Z,2026-09-17T13:02:49Z,0,0,UCHhc-QgZAAVrzXL_ILpmYmQ,2026-09-18T08:12:52.586097+00:00,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,14,uncertain
3,R4-zhQcOecM,UgzveMhZ_NaaGxXCiFR4AaABAg,NaN,False,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",2026-09-17T11:45:19Z,2026-09-17T11:45:19Z,0,0,UCa7-Fz3J4jsnMdQ-P1L14RA,2026-09-18T08:12:52.586097+00:00,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",18,uncertain
4,R4-zhQcOecM,UgyMy7oNdvD2Opwlh8d4AaABAg,NaN,False,nice,2026-09-17T08:27:14Z,2026-09-17T08:27:14Z,0,0,UC7KFZSWH82DKaNH2uCT5j8w,2026-09-18T08:12:52.586097+00:00,nice,1,uncertain
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24343,lQRIvQ28wVA,UgxjH9-MKlPO1ZyR9Y94AaABAg,NaN,False,Nice place!😍,2021-12-25 14:11:01+00:00,2021-12-25T14:11:01Z,1,0,UC5aczqXyfmQk2k0GBxb9ayg,2026-09-18T09:17:49.905232+00:00,Nice place!😍,2,uncertain
24344,lQRIvQ28wVA,Ugz6CMvXjElVk_3Fks94AaABAg,NaN,False,"super uplaod, Joel. )",2021-12-23 09:48:48+00:00,2021-12-23T09:48:48Z,2,1,UCsrDZMfP9xvY374d8RMf2Kw,2026-09-18T09:17:49.905257+00:00,"super uplaod, Joel. )",3,uncertain
24345,lQRIvQ28wVA,UgyBSWRWA8n8MdmgG1N4AaABAg,NaN,False,A wonderful MALAYSIA!,2021-12-23 09:19:56+00:00,2021-12-23T09:19:56Z,2,1,UCWeXFP1x6lsGeKGhDHgkM5w,2026-09-18T09:17:49.905267+00:00,A wonderful MALAYSIA!,3,tourism_relevant
24346,lQRIvQ28wVA,UgyoYTpR0e4Bi3K0jqF4AaABAg,NaN,False,Nice vlog dude.,2021-12-23 09:18:36+00:00,2021-12-23T09:18:36Z,2,1,UCoEQik-7LIBrvKSs0Ju9aIw,2026-09-18T09:17:49.905276+00:00,Nice vlog dude.,3,tourism_relevant


In [16]:
df["relevance"].value_counts()

,count
relevance,
tourism_relevant,10805
uncertain,10165
sentiment_only,2056
irrelevant_noise,363


In [17]:
df[df["relevance"] == "tourism_relevant"][["clean_text", "relevance"]].head(30)

,clean_text,relevance
88,"Anh đăng video 3 ngày được 1,05 triệu view lun...",tourism_relevant
89,"Không biết thương về Việt Nam chưa ? , Nêu Thư...",tourism_relevant
98,Nhìn anh em tưởng đâu ạ là em của lâm vlog vid...,tourism_relevant
100,Ở malaysia khi đi ăn những chỗ này thì mình tự...,tourism_relevant
141,Vaiz hôm bữa em cũng đi với các bạn đến malays...,tourism_relevant
150,anh thử review quán ăn ở đức đi ạ,tourism_relevant
158,Malai có rất nhiều Chinese restaurants ngon gh...,tourism_relevant
163,Cái này quán food halal,tourism_relevant
182,Bên Malay họ ăn nhiều nên là 1 phần ăn nó to l...,tourism_relevant
270,"Trung Quốc nha a, TQ rộng lớn có nhiều món rất...",tourism_relevant


In [18]:
df[df["relevance"] == "irrelevant_noise"][["clean_text", "relevance"]].head(30)

,clean_text,relevance
7,❤❤❤❤❤,irrelevant_noise
26,🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣,irrelevant_noise
37,❤❤❤❤❤❤❤😊🎉,irrelevant_noise
138,❤❤❤❤,irrelevant_noise
192,hello🥰🥰🥰,irrelevant_noise
262,​@heyhohayquađúng 😊,irrelevant_noise
327,​@UtnhoUtnho-j1hkaiziwkzisz,irrelevant_noise
375,🅸︎🅰︎🅼︎ 🅼︎🅰︎🅻︎🅰︎🆈︎🆂︎🅸︎🅰︎,irrelevant_noise
676,🤩🤩🤩,irrelevant_noise
750,👍👏👏👏❤️❤️❤️🇲🇾🇲🇾🇲🇾,irrelevant_noise


In [19]:
df[df["relevance"] == "uncertain"][["clean_text", "relevance"]].head(50)

,clean_text,relevance
0,Cho mình biết bạn muốn mình đến quán bẫy du lị...,uncertain
1,"Mong anh làm khách sạn ma ám ở VN ạ, mong lên top",uncertain
2,Ngon quá xấu chỗ này chắc mai đi bây giờ thứ 2...,uncertain
3,"Anh ra video tiếp đi ạ, ny em ăn cơm phải coi ...",uncertain
4,nice,uncertain
5,Hay quá,uncertain
6,Ủng hộ anh,uncertain
8,Đi du lịch Nhật Bản đi Thương,uncertain
9,Lần sau mk sẽ vẽ ông người nhật nha các bạn 🎉🎉🎉,uncertain
10,Ra thêm nhiu clip nhoa❤❤,uncertain


In [20]:
df.to_csv("v1.csv", index=False)

In [21]:
!pip install lingua-language-detector

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.3/170.3 MB 7.4 MB/s eta 0:00:00


In [22]:
from lingua import LanguageDetectorBuilder

detector = (
    LanguageDetectorBuilder
    .from_all_languages()
    .build()
)

def detect_language(text):

    try:

        result = detector.detect_language_of(
            str(text)
        )

        if result:
            return result.iso_code_639_1.name

        return "unknown"

    except:
        return "unknown"


df["language"] = (
    df["clean_text"]
    .apply(detect_language)
)

In [23]:
print(
    df["language"]
    .value_counts()
    .head(20)
)

language
EN         17012
MS          1746
ID          1021
TL           647
VI           318
unknown      307
SW           171
LA           157
YO           125
SO           101
SN            99
ZH            95
ST            91
TS            85
TN            83
NL            76
XH            69
DE            65
FR            65
AF            63
Name: count, dtype: int64


In [24]:
import pandas as pd
import numpy as np
import re

# ============================================================
# 1. LOAD DATA
# ============================================================

INPUT_FILE = "v1.csv"

df = pd.read_csv(INPUT_FILE)

print("Original dataset:", df.shape)
print(df.columns.tolist())


# ============================================================
# 2. BASIC CLEANING
# ============================================================

df["clean_text"] = (
    df["clean_text"]
    .fillna("")
    .astype(str)
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

df["comment_text"] = (
    df["comment_text"]
    .fillna("")
    .astype(str)
)


# ============================================================
# 3. BETTER WORD COUNT
# ============================================================
# Counts actual text-like words rather than emoji chunks.

df["word_count_new"] = (
    df["clean_text"]
    .str.findall(r"\b[\w']+\b")
    .str.len()
)


# ============================================================
# 4. HELPER FUNCTIONS
# ============================================================

def normalize(text):
    """
    Lowercase + normalize whitespace.
    """
    text = str(text).lower()
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def contains_any(text, keywords):
    """
    Returns True if any keyword exists in text.
    """
    text = normalize(text)

    return any(keyword.lower() in text for keyword in keywords)


def count_keywords(text, keywords):
    """
    Count how many keywords occur.
    """
    text = normalize(text)

    return sum(
        1 for keyword in keywords
        if keyword.lower() in text
    )


# ============================================================
# 5. OBVIOUS NOISE
# ============================================================

NOISE_PHRASES = {
    "first",
    "second",
    "third",
    "thanks",
    "thank you",
    "thank u",
    "tq",
    "tqtq",
    "hi",
    "hello",
    "lol",
    "lmao",
    "haha",
    "hahaha",
    "nice video",
    "great video",
    "good video",
    "love your video",
    "love this video",
    "new subscriber",
    "subscribed",
}


def is_noise(text):

    text = normalize(text)

    # Empty
    if text == "":
        return True

    # Remove @mentions
    no_mentions = re.sub(r"@\S+", "", text).strip()

    # If only emoji/punctuation/symbols remain
    if not re.search(r"[a-zA-Z0-9À-ÿ]", no_mentions):
        return True

    # Simplified text
    simple = re.sub(
        r"[^a-zA-Z0-9À-ÿ\s']",
        "",
        no_mentions
    )

    simple = re.sub(r"\s+", " ", simple).strip()

    if simple in NOISE_PHRASES:
        return True

    return False


df["is_noise"] = df["clean_text"].apply(is_noise)


# ============================================================
# 6. DETECT COMMENT TARGET
# ============================================================
#
# We want to distinguish:
#
# destination -> about Malaysia/place/travel experience
# creator     -> about YouTuber/video/editing/person
# unrelated   -> obviously unrelated
# unknown     -> not enough evidence


CREATOR_KEYWORDS = [
    "your video",
    "great video",
    "nice video",
    "good video",
    "amazing video",
    "love your videos",
    "love your vlog",
    "editing",
    "transition",
    "camera",
    "iphone",
    "subscriber",
    "subscribe",
    "channel",
    "content",
    "youtuber",
    "vlogger",
    "your videos",
    "next video",
    "new video",
    "beautiful video",
    "great job bro",
    "great job sis",
    "you guys",
    "your family",
]


TOURISM_GENERAL = [
    "travel",
    "travelling",
    "traveling",
    "traveler",
    "traveller",
    "tourist",
    "tourism",
    "trip",
    "vacation",
    "holiday",
    "visit",
    "visited",
    "visiting",
    "destination",
    "itinerary",
    "backpacking",
    "backpacker",
    "sightseeing",
]


MALAYSIA_TERMS = [
    "malaysia",
    "malaysian",
    "kl",
    "kuala lumpur",
    "penang",
    "georgetown",
    "george town",
    "langkawi",
    "malacca",
    "melaka",
    "johor",
    "johor bahru",
    "sabah",
    "sarawak",
    "borneo",
    "putrajaya",
    "ipoh",
    "perak",
    "selangor",
    "terengganu",
    "kelantan",
    "pahang",
    "kedah",
    "perlis",
    "negeri sembilan",
    "tioman",
    "perhentian",
    "redang",
    "semporna",
    "sipadan",
    "kinabalu",
    "kota kinabalu",
    "kuching",
    "genting",
    "cameron highlands",
]


TOURISM_EXPERIENCE_TERMS = [
    "hotel",
    "hostel",
    "resort",
    "beach",
    "island",
    "airport",
    "flight",
    "train",
    "bus",
    "grab",
    "taxi",
    "food",
    "restaurant",
    "street food",
    "hawker",
    "market",
    "museum",
    "temple",
    "mosque",
    "park",
    "waterfall",
    "mountain",
    "diving",
    "snorkeling",
    "scuba",
    "tour guide",
    "ticket",
    "entrance fee",
]


def identify_target(text):

    text = normalize(text)

    if is_noise(text):
        return "noise"

    creator_score = count_keywords(
        text,
        CREATOR_KEYWORDS
    )

    destination_score = (
        count_keywords(text, TOURISM_GENERAL)
        + count_keywords(text, MALAYSIA_TERMS)
        + count_keywords(text, TOURISM_EXPERIENCE_TERMS)
    )

    # Strong creator signal but no destination discussion
    if creator_score > 0 and destination_score == 0:
        return "creator"

    if destination_score > 0:
        return "destination"

    return "unknown"


df["target"] = df["clean_text"].apply(
    identify_target
)


# ============================================================
# 7. STAKEHOLDER IDENTIFICATION
# ============================================================
#
# IMPORTANT:
# Only classify LOCAL/VISITOR where the commenter explicitly
# provides evidence.
#
# Do NOT infer nationality from username, language, etc.


LOCAL_PATTERNS = [
    r"\bas a malaysian\b",
    r"\bi am malaysian\b",
    r"\bi'm malaysian\b",
    r"\bim malaysian\b",
    r"\bmalaysian here\b",
    r"\bas a local\b",
    r"\bi am a local\b",
    r"\bi'm a local\b",
    r"\blocal here\b",
    r"\bi live in malaysia\b",
    r"\bi live in kl\b",
    r"\bi live in kuala lumpur\b",
    r"\bborn and raised in malaysia\b",
    r"\bborn and raised in kl\b",
    r"\bfrom malaysia\b",
]


VISITOR_PATTERNS = [
    r"\bi visited malaysia\b",
    r"\bvisited malaysia\b",
    r"\bwhen i visited\b",
    r"\bi went to malaysia\b",
    r"\btravelled to malaysia\b",
    r"\btraveled to malaysia\b",
    r"\bmy trip to malaysia\b",
    r"\bmy malaysia trip\b",
    r"\bspent .* days in malaysia\b",
    r"\bspent .* days in kl\b",
    r"\bspent .* days in kuala lumpur\b",
    r"\bwhen i was in malaysia\b",
    r"\bwhen we were in malaysia\b",
    r"\bi stayed in malaysia\b",
    r"\bwe stayed in malaysia\b",
    r"\bi came to malaysia\b",
]


FUTURE_VISITOR_PATTERNS = [
    r"\bi want to visit\b",
    r"\bi want to go\b",
    r"\bplanning to visit\b",
    r"\bplanning a trip\b",
    r"\bwill visit\b",
    r"\bgoing to malaysia\b",
    r"\bvisit malaysia someday\b",
    r"\bmy bucket list\b",
]


def stakeholder(text):

    text = normalize(text)

    for pattern in LOCAL_PATTERNS:
        if re.search(pattern, text):
            return "local"

    for pattern in VISITOR_PATTERNS:
        if re.search(pattern, text):
            return "visitor"

    for pattern in FUTURE_VISITOR_PATTERNS:
        if re.search(pattern, text):
            return "potential_visitor"

    return "unknown"


df["stakeholder"] = df["clean_text"].apply(
    stakeholder
)


# ============================================================
# 8. TOURISM / SUSTAINABILITY ASPECTS
# ============================================================

ASPECT_KEYWORDS = {

    # ----------------------------
    # ENVIRONMENT
    # ----------------------------

    "cleanliness": [
        "clean",
        "dirty",
        "rubbish",
        "trash",
        "garbage",
        "litter",
        "waste",
        "plastic",
        "smelly",
        "polluted",
        "pollution",
    ],

    "beach_marine": [
        "beach",
        "sea",
        "ocean",
        "coast",
        "coastal",
        "coral",
        "reef",
        "marine",
        "diving",
        "snorkeling",
        "water quality",
    ],

    "nature_wildlife": [
        "nature",
        "forest",
        "jungle",
        "wildlife",
        "animal",
        "monkey",
        "turtle",
        "orangutan",
        "national park",
        "mountain",
        "waterfall",
        "biodiversity",
    ],

    "climate_weather": [
        "weather",
        "rain",
        "rainy",
        "monsoon",
        "flood",
        "flooding",
        "hot",
        "heat",
        "humid",
        "climate",
    ],


    # ----------------------------
    # SOCIAL
    # ----------------------------

    "crowding": [
        "crowded",
        "crowd",
        "too many tourists",
        "overtourism",
        "busy",
        "packed",
        "queue",
        "long queue",
        "too many people",
    ],

    "safety": [
        "safe",
        "unsafe",
        "dangerous",
        "crime",
        "scam",
        "scammer",
        "pickpocket",
        "robbery",
        "security",
    ],

    "hospitality": [
        "friendly",
        "helpful",
        "kind people",
        "nice people",
        "locals",
        "local people",
        "hospitality",
        "welcoming",
        "rude",
    ],

    "culture_heritage": [
        "culture",
        "cultural",
        "heritage",
        "traditional",
        "tradition",
        "temple",
        "mosque",
        "historic",
        "history",
        "local culture",
        "authentic",
    ],


    # ----------------------------
    # ECONOMIC
    # ----------------------------

    "price_value": [
        "cheap",
        "expensive",
        "price",
        "prices",
        "cost",
        "costly",
        "affordable",
        "overpriced",
        "value for money",
        "ringgit",
        "rm ",
    ],

    "local_business": [
        "local business",
        "small business",
        "hawker",
        "market",
        "vendor",
        "seller",
        "shop",
        "restaurant",
        "local restaurant",
    ],

    "food": [
        "food",
        "nasi lemak",
        "laksa",
        "satay",
        "char kway teow",
        "roti canai",
        "restaurant",
        "hawker",
        "street food",
        "delicious",
        "sedap",
        "makanan",
    ],


    # ----------------------------
    # INFRASTRUCTURE
    # ----------------------------

    "transport": [
        "public transport",
        "transport",
        "train",
        "lrt",
        "mrt",
        "monorail",
        "bus",
        "grab",
        "taxi",
        "airport",
        "flight",
        "traffic",
        "road",
        "parking",
    ],

    "accessibility": [
        "accessible",
        "accessibility",
        "wheelchair",
        "disabled",
        "elderly",
        "walking",
        "walkable",
        "pedestrian",
        "sidewalk",
    ],

    "facilities": [
        "toilet",
        "toilets",
        "restroom",
        "facility",
        "facilities",
        "signage",
        "sign",
        "wifi",
        "internet",
        "parking",
    ],

    "accommodation": [
        "hotel",
        "hostel",
        "resort",
        "airbnb",
        "accommodation",
        "room",
        "stay",
        "homestay",
    ],

    "seasonality": [
        "school holiday",
        "school holidays",
        "public holiday",
        "public holidays",
        "peak season",
        "off season",
        "high season",
        "low season",
        "monsoon season",
        "holiday season",
    ],
}


def detect_aspects(text):

    text = normalize(text)

    found = []

    for aspect, keywords in ASPECT_KEYWORDS.items():

        if contains_any(text, keywords):
            found.append(aspect)

    if not found:
        return "none"

    return ";".join(found)


df["tourism_aspect"] = df["clean_text"].apply(
    detect_aspects
)


# ============================================================
# 9. MAP ASPECTS -> SUSTAINABILITY DIMENSIONS
# ============================================================

DIMENSION_MAP = {

    "cleanliness": "environmental",
    "beach_marine": "environmental",
    "nature_wildlife": "environmental",
    "climate_weather": "environmental",

    "crowding": "social",
    "safety": "social",
    "hospitality": "social",
    "culture_heritage": "social",

    "price_value": "economic",
    "local_business": "economic",
    "food": "economic",

    "transport": "infrastructure",
    "accessibility": "infrastructure",
    "facilities": "infrastructure",
    "accommodation": "infrastructure",
    "seasonality": "infrastructure",
}


def get_dimensions(aspects):

    if aspects == "none":
        return "none"

    dimensions = set()

    for aspect in aspects.split(";"):

        if aspect in DIMENSION_MAP:
            dimensions.add(
                DIMENSION_MAP[aspect]
            )

    if not dimensions:
        return "none"

    return ";".join(sorted(dimensions))


df["sustainability_dimension"] = (
    df["tourism_aspect"]
    .apply(get_dimensions)
)


# ============================================================
# 10. DESTINATION EXTRACTION
# ============================================================

DESTINATIONS = {

    "Kuala Lumpur": [
        "kuala lumpur",
        " kl ",
        " kl.",
    ],

    "Penang": [
        "penang",
        "georgetown",
        "george town",
    ],

    "Langkawi": [
        "langkawi",
    ],

    "Melaka": [
        "melaka",
        "malacca",
    ],

    "Johor": [
        "johor",
        "johor bahru",
        "jb",
    ],

    "Sabah": [
        "sabah",
        "kota kinabalu",
        "kinabalu",
        "semporna",
        "sipadan",
    ],

    "Sarawak": [
        "sarawak",
        "kuching",
    ],

    "Putrajaya": [
        "putrajaya",
    ],

    "Perak": [
        "perak",
        "ipoh",
    ],

    "Pahang": [
        "pahang",
        "cameron highlands",
        "genting",
    ],

    "Terengganu": [
        "terengganu",
        "redang",
        "perhentian",
    ],

    "Malaysia": [
        "malaysia",
        "malaysian",
    ],
}


def detect_destination(text):

    text = " " + normalize(text) + " "

    found = []

    for destination, terms in DESTINATIONS.items():

        if any(term in text for term in terms):
            found.append(destination)

    # If specific state/place found, Malaysia doesn't need
    # to appear separately
    if len(found) > 1 and "Malaysia" in found:
        found.remove("Malaysia")

    if not found:
        return "unknown"

    return ";".join(found)


df["destination"] = df["clean_text"].apply(
    detect_destination
)


# ============================================================
# 11. SIMPLE RULE-BASED SENTIMENT
# ============================================================
#
# This is ONLY a first-pass sentiment system.
# Later you can replace this with a multilingual transformer.


POSITIVE_WORDS = [
    "good",
    "great",
    "amazing",
    "beautiful",
    "awesome",
    "excellent",
    "love",
    "loved",
    "like",
    "best",
    "perfect",
    "friendly",
    "helpful",
    "clean",
    "safe",
    "cheap",
    "affordable",
    "delicious",
    "nice",
    "wonderful",
    "fantastic",
    "enjoy",
    "enjoyed",
    "recommend",
    "recommended",
    "worth",
    "underrated",
    "cantik",
    "cantek",
    "sedap",
]


NEGATIVE_WORDS = [
    "bad",
    "terrible",
    "awful",
    "hate",
    "worst",
    "dirty",
    "unsafe",
    "dangerous",
    "expensive",
    "overpriced",
    "crowded",
    "polluted",
    "pollution",
    "rubbish",
    "garbage",
    "trash",
    "scam",
    "scammer",
    "disappointed",
    "disappointing",
    "boring",
    "poor",
    "problem",
    "traffic",
]


def simple_sentiment(text):

    text = normalize(text)

    positive = count_keywords(
        text,
        POSITIVE_WORDS
    )

    negative = count_keywords(
        text,
        NEGATIVE_WORDS
    )

    if positive > negative:
        return "positive"

    elif negative > positive:
        return "negative"

    else:
        return "neutral"


df["sentiment"] = df["clean_text"].apply(
    simple_sentiment
)


# ============================================================
# 12. HELPFULNESS SCORE 0 - 5
# ============================================================
#
# 0 = useless/noise
# 1 = very low information
# 2 = general tourism opinion
# 3 = specific tourism information
# 4 = useful sustainability observation
# 5 = highly informative / actionable


EXPERIENCE_PATTERNS = [
    r"\bi visited\b",
    r"\bwe visited\b",
    r"\bi went\b",
    r"\bwe went\b",
    r"\bi stayed\b",
    r"\bwe stayed\b",
    r"\bi travelled\b",
    r"\bi traveled\b",
    r"\bwhen i was\b",
    r"\bwhen we were\b",
    r"\bspent .* day",
    r"\bmy experience\b",
    r"\bin my experience\b",
]


ACTIONABLE_WORDS = [
    "should",
    "need to",
    "needs to",
    "must",
    "recommend",
    "avoid",
    "better",
    "improve",
    "problem",
    "issue",
    "too many",
    "too much",
    "not enough",
    "please",
    "tip",
    "advice",
]


def helpfulness_score(row):

    text = normalize(row["clean_text"])

    # ------------------------
    # Score 0
    # ------------------------

    if row["is_noise"]:
        return 0

    if row["target"] == "creator":
        return 0

    score = 1


    # ------------------------
    # Tourism context
    # ------------------------

    if row["target"] == "destination":
        score += 1


    # ------------------------
    # Contains a tourism aspect
    # ------------------------

    if row["tourism_aspect"] != "none":
        score += 1


    # ------------------------
    # First-hand experience
    # ------------------------

    if any(
        re.search(pattern, text)
        for pattern in EXPERIENCE_PATTERNS
    ):
        score += 1


    # ------------------------
    # Specific actionable information
    # ------------------------

    if contains_any(
        text,
        ACTIONABLE_WORDS
    ):
        score += 1


    # ------------------------
    # Detailed sustainability comment
    # ------------------------

    if (
        row["sustainability_dimension"] != "none"
        and row["word_count_new"] >= 10
    ):
        score += 1


    # ------------------------
    # Very short comments cannot be
    # considered highly informative
    # ------------------------

    if row["word_count_new"] <= 2:
        score = min(score, 2)


    # Max 5
    return min(score, 5)


df["helpfulness_score"] = df.apply(
    helpfulness_score,
    axis=1
)


# ============================================================
# 13. CREATE ANALYSIS FLAGS
# ============================================================

# Broad tourism comments
df["tourism_useful"] = (
    (df["target"] == "destination")
    &
    (df["helpfulness_score"] >= 2)
)


# Strong comments for analysis
df["analysis_ready"] = (
    (df["target"] == "destination")
    &
    (df["sustainability_dimension"] != "none")
    &
    (df["helpfulness_score"] >= 3)
)


# Very strong / actionable comments
df["high_value"] = (
    df["analysis_ready"]
    &
    (df["helpfulness_score"] >= 4)
)


# ============================================================
# 14. WEIGHT USING LIKES
# ============================================================
#
# Do NOT let one viral comment dominate everything.
# log1p compresses large like counts.

df["like_count"] = pd.to_numeric(
    df["like_count"],
    errors="coerce"
).fillna(0)

df["engagement_weight"] = np.log1p(
    df["like_count"]
)


# ============================================================
# 15. EXPORT DATASETS
# ============================================================

# Complete classified dataset
df.to_csv(
    "comments_classified.csv",
    index=False
)


# Tourism-related useful comments
tourism_df = df[
    df["tourism_useful"]
].copy()

tourism_df.to_csv(
    "tourism_useful_comments.csv",
    index=False
)


# Sustainability dataset
sustainable_df = df[
    df["analysis_ready"]
].copy()

sustainable_df.to_csv(
    "sustainable_tourism_comments.csv",
    index=False
)


# Highest-value comments
high_value_df = df[
    df["high_value"]
].copy()

high_value_df.to_csv(
    "high_value_comments.csv",
    index=False
)


# Malaysian/local comments
local_df = sustainable_df[
    sustainable_df["stakeholder"] == "local"
].copy()

local_df.to_csv(
    "local_sustainability_comments.csv",
    index=False
)


# Visitor comments
visitor_df = sustainable_df[
    sustainable_df["stakeholder"] == "visitor"
].copy()

visitor_df.to_csv(
    "visitor_sustainability_comments.csv",
    index=False
)


# ============================================================
# 16. SUMMARY
# ============================================================

print("\n" + "=" * 70)
print("PROCESSING COMPLETE")
print("=" * 70)

print("\nTotal comments:")
print(len(df))

print("\nTARGET:")
print(df["target"].value_counts())

print("\nSTAKEHOLDER:")
print(df["stakeholder"].value_counts())

print("\nHELPFULNESS:")
print(
    df["helpfulness_score"]
    .value_counts()
    .sort_index()
)

print("\nSUSTAINABILITY DIMENSION:")
print(
    df["sustainability_dimension"]
    .value_counts()
    .head(20)
)

print("\nSENTIMENT:")
print(
    sustainable_df["sentiment"]
    .value_counts()
)

print("\nTourism useful:")
print(len(tourism_df))

print("\nSustainability analysis ready:")
print(len(sustainable_df))

print("\nHigh-value comments:")
print(len(high_value_df))

print("\nExplicit local comments:")
print(len(local_df))

print("\nExplicit visitor comments:")
print(len(visitor_df))


# ============================================================
# 17. SHOW HIGH-VALUE EXAMPLES
# ============================================================

columns_to_show = [
    "clean_text",
    "stakeholder",
    "destination",
    "tourism_aspect",
    "sustainability_dimension",
    "sentiment",
    "helpfulness_score",
    "like_count"
]

print("\n" + "=" * 70)
print("HIGH-VALUE EXAMPLES")
print("=" * 70)

print(
    high_value_df[
        columns_to_show
    ]
    .sort_values(
        ["helpfulness_score", "like_count"],
        ascending=False
    )
    .head(30)
    .to_string(index=False)
)

Original dataset: (23389, 14)
['video_id', 'comment_id', 'parent_comment_id', 'is_reply', 'comment_text', 'comment_published_at', 'comment_updated_at', 'like_count', 'reply_count', 'author_channel_id', 'collected_at', 'clean_text', 'word_count', 'relevance']

PROCESSING COMPLETE

Total comments:
23389

TARGET:
target
unknown        11146
destination    10624
creator          948
noise            671
Name: count, dtype: int64

STAKEHOLDER:
stakeholder
unknown              22742
local                  524
potential_visitor       84
visitor                 39
Name: count, dtype: int64

HELPFULNESS:
helpfulness_score
0    1619
1    9217
2    6216
3    1916
4    3171
5    1250
Name: count, dtype: int64

SUSTAINABILITY DIMENSION:
sustainability_dimension
none                                            17259
economic                                         1789
infrastructure                                   1054
environmental                                     919
social                   

In [25]:
df.to_csv("final.csv", index=False)

---

# Sentiment Modelling & Dashboard-Ready Aggregation

**Data quality notes before extending the pipeline** (two things worth a deliberate decision, not a silent fix):

1. **The "exclude 2026" date filter never actually ran.** Cells 4 and 5 above (the `comment_published_at` filtering
   logic) are formatted as **Markdown**, not **Code** — so they were never executed, and `final.csv` was never
   filtered to `< 2026-01-01`. As a result, **3,968 of 23,389 comments (17%) are dated in 2026**, including
   comments up to the day this data was scraped. This may well be *correct* — `collected_at` shows the scrape
   happened today, so these look like genuine recent comments, and force-applying the old cutoff would delete
   real data. But since it's not clear whether the `< 2026-01-01` filter was a deliberate scope decision or a
   leftover from an earlier version of the collector, **this is flagged rather than silently changed** — if you
   want that filter applied, convert cells 4–5 to Code and re-run from `merged_comments.csv`.
2. **Language detection (cells 23–25) never made it into `final.csv`.** Cell 26 reloads `df` fresh from `v1.csv`
   (`df = pd.read_csv("v1.csv")`), which discards the in-memory `language` column added in cell 24 before the
   classification pipeline runs — so `final.csv` has no `language` column at all. Since the suggestion below
   depends on knowing which comments are English vs not (for TourBERT vs XLM-R routing), **Stage 1 below
   recomputes language detection**, scoped only to the 4,755 `analysis_ready` comments (~13s) rather than all
   23,389, so it doesn't silently disappear again.

---

## Why this extends rather than replaces `sentiment`

The existing `sentiment` column (keyword-based: more positive words than negative words → positive) is a
reasonable first pass, but two things limit it: it can't handle mixed sentiment in one comment (*"food is good, but
country is fucking suck"* currently scores **positive**, purely because "good" is present), and it doesn't
generalize to the ~8% of `analysis_ready` comments that aren't in English. The plan below keeps `sentiment` intact
as a labelled baseline (renamed `rule_sentiment`), adds a multilingual transformer baseline that needs no manual
labelling, and gates a fine-tuned tourism-domain model behind actually having gold labels to fine-tune on —
each stage is checked against this data rather than assumed to work.

## Stage 1 — Prepare the sentiment-ready dataset

Scope to `analysis_ready == True` (the 4,755 comments that are on-topic, tourism-relevant, and have a
sustainability dimension attached — already filtered upstream). Preserve the existing rule-based label under a
new name so later stages can be compared against it rather than overwriting it.

In [26]:
import pandas as pd
import numpy as np

df = pd.read_csv("final.csv")

sentiment_df = df[df["analysis_ready"] == True].copy().reset_index(drop=True)
sentiment_df = sentiment_df.rename(columns={"sentiment": "rule_sentiment"})

print("analysis_ready comments:", len(sentiment_df))
print("\nrule_sentiment baseline (treat as OLD baseline, not ground truth):")
counts = sentiment_df["rule_sentiment"].value_counts()
pct = (counts / counts.sum() * 100).round(1)
print(pd.DataFrame({"count": counts, "pct": pct}))


analysis_ready comments: 4755

rule_sentiment baseline (treat as OLD baseline, not ground truth):
                count   pct
rule_sentiment             
positive         2351  49.4
neutral          2061  43.3
negative          343   7.2


In [27]:
# Recompute language detection on the WORKING SET ONLY (13s vs minutes on all 23k rows) -
# the version in cells 23-25 above was silently dropped when cell 26 reloaded df from v1.csv.
try:
    from lingua import LanguageDetectorBuilder
except ImportError:
    import sys, subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "lingua-language-detector", "-q"])
    from lingua import LanguageDetectorBuilder

detector = LanguageDetectorBuilder.from_all_languages().build()

def detect_language(text):
    try:
        result = detector.detect_language_of(str(text))
        return result.iso_code_639_1.name if result else "unknown"
    except Exception:
        return "unknown"

sentiment_df["language"] = sentiment_df["clean_text"].apply(detect_language)

# Coarse EN / non-EN routing flag. Fine-grained codes on very short comments are noisy
# (lingua misfires on a handful of words) - what actually matters for TourBERT-vs-XLM-R
# routing is just "is this English or not".
sentiment_df["is_english"] = sentiment_df["language"].eq("EN")

print(sentiment_df["language"].value_counts().head(10))
print(f"\nEnglish share: {sentiment_df['is_english'].mean():.1%} "
      f"({sentiment_df['is_english'].sum()} of {len(sentiment_df)})")


language
EN    4369
MS     130
ID     128
TL      77
LA       7
VI       5
ST       4
SN       4
YO       4
EO       3
Name: count, dtype: int64

English share: 91.9% (4369 of 4755)


## Stage 2 — Multilingual pretrained sentiment baseline (XLM-R)

**Requires internet access to huggingface.co and, ideally, a GPU — run this in Colab, not in an offline/sandboxed
environment.** `cardiffnlp/twitter-xlm-roberta-base-sentiment` needs no fine-tuning and covers every language in
this dataset in one pass, so it becomes the baseline every other model is compared against. The result is cached
to `sentiment_xlmr.csv` so re-running the notebook doesn't re-download/re-infer unnecessarily.

In [28]:
import os

XLMR_CACHE = "sentiment_xlmr.csv"
MODEL_NAME = "cardiffnlp/twitter-xlm-roberta-base-sentiment"

if os.path.exists(XLMR_CACHE):
    print(f"Loading cached XLM-R predictions from {XLMR_CACHE} (delete this file to force re-inference).")
    xlmr_predictions = pd.read_csv(XLMR_CACHE)
    assert len(xlmr_predictions) == len(sentiment_df), (
        "Cached predictions don't match the current sentiment_df length - delete the cache and re-run."
    )
else:
    import torch
    from transformers import AutoTokenizer, AutoModelForSequenceClassification
    from tqdm.auto import tqdm

    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
    model = AutoModelForSequenceClassification.from_pretrained(MODEL_NAME)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)
    model.eval()
    print("Using:", device, "- inference will be slow on CPU for 4,755 comments; a Colab GPU runtime is recommended.")

    def predict_sentiment_batch(texts, batch_size=32):
        results = []
        label_names = ["negative", "neutral", "positive"]
        for i in tqdm(range(0, len(texts), batch_size)):
            batch = texts[i:i + batch_size]
            encoded = tokenizer(batch, padding=True, truncation=True, max_length=128, return_tensors="pt")
            encoded = {k: v.to(device) for k, v in encoded.items()}
            with torch.no_grad():
                output = model(**encoded)
            probs = torch.softmax(output.logits, dim=1).cpu().numpy()
            for p in probs:
                results.append({
                    "model_sentiment": label_names[p.argmax()],
                    "sentiment_confidence": float(p.max()),
                    "negative_probability": float(p[0]),
                    "neutral_probability": float(p[1]),
                    "positive_probability": float(p[2]),
                })
        return pd.DataFrame(results)

    texts = sentiment_df["clean_text"].fillna("").astype(str).tolist()
    xlmr_predictions = predict_sentiment_batch(texts)
    xlmr_predictions.to_csv(XLMR_CACHE, index=False)
    print(f"Saved predictions to {XLMR_CACHE}")

xlmr_predictions.index = sentiment_df.index
sentiment_df = pd.concat(
    [sentiment_df.drop(columns=xlmr_predictions.columns, errors="ignore"), xlmr_predictions],
    axis=1
)
sentiment_df[["clean_text", "rule_sentiment", "model_sentiment", "sentiment_confidence"]].head(10)


config.json:   0%|          | 0.00/841 [00:00<?, ?B/s]

sentencepiece.bpe.model: reconstructing file:   0%|          |  0.00B / 5.07MB            

sentencepiece.bpe.model: downloading bytes:           |  0.00B            

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

pytorch_model.bin: reconstructing file:   0%|          |  0.00B / 1.11GB            

pytorch_model.bin: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

Using: cpu - inference will be slow on CPU for 4,755 comments; a Colab GPU runtime is recommended.


  0%|          | 0/149 [00:00<?, ?it/s]

model.safetensors: reconstructing file:   0%|          |  0.00B / 1.11GB            

model.safetensors: downloading bytes:           |  0.00B            

Saved predictions to sentiment_xlmr.csv


,clean_text,rule_sentiment,model_sentiment,sentiment_confidence
0,Malai có rất nhiều Chinese restaurants ngon gh...,neutral,negative,0.379045
1,Cái này quán food halal,neutral,neutral,0.627348
2,"Malaysia food is good, but country is fucking ...",positive,negative,0.937937
3,I'm Malaysian I will tell you how it's like: 1...,positive,negative,0.594075
4,Try nasi lemak the most popular food from Mala...,neutral,positive,0.488678
5,Try some nasi lemak and roti canai they are th...,positive,positive,0.568097
6,we actually have the grab motorcycle idk why t...,neutral,negative,0.527481
7,Ok you said what is RM Stands for RM stands fo...,neutral,positive,0.556861
8,"When I was a little,my mom and dad said ""if ur...",neutral,neutral,0.511060
9,We actually got GRAB here in Malaysia tho 😭,neutral,negative,0.840825


**Why probabilities are stored, not just the label:** *"Beach is dirty"* at `negative_probability=0.91` and *"Food
is okay, but nothing amazing"* at `neutral_probability=0.71`/`positive_probability=0.17` both currently collapse to
one word each under the old rule system. Keeping all three probabilities plus `sentiment_confidence` lets the
dashboard distinguish **strong** sentiment from **uncertain** sentiment — e.g. only surfacing a comment under "Top
Negative Issues" when `negative_probability` clears a threshold, not just when it happens to win a 3-way tie.

In [29]:
# Sanity check with NO gold labels needed yet: how often do the rule-based baseline
# and the XLM-R baseline agree? Large disagreement is expected and informative, not a bug -
# it's exactly the mixed-sentiment / non-English cases the rule system couldn't handle.
agreement = (sentiment_df["rule_sentiment"] == sentiment_df["model_sentiment"]).mean()
print(f"Rule-based vs XLM-R agreement: {agreement:.1%}\n")
print("Confusion matrix (rows = rule_sentiment, columns = model_sentiment):")
print(pd.crosstab(sentiment_df["rule_sentiment"], sentiment_df["model_sentiment"]))


Rule-based vs XLM-R agreement: 54.4%

Confusion matrix (rows = rule_sentiment, columns = model_sentiment):
model_sentiment  negative  neutral  positive
rule_sentiment                              
negative              311       17        15
neutral               737      749       575
positive              564      262      1525


## Stage 3 — Build a gold-standard annotation sample (stratified, not naive random)

A plain random sample of 1,000 comments would inherit the rule-based class imbalance almost exactly
(~49% positive / ~43% neutral / ~7% negative, per Stage 1), leaving only ~72 negative examples to evaluate against
— too few to trust a negative-class F1 score. Since negative comments are exactly the sustainability problems this
project needs to detect, the sample below **keeps every rule-flagged negative comment** and fills the rest evenly
from positive/neutral, so all three classes have a comparable number of examples to hand-label and evaluate on.

In [30]:
TARGET_N = 1000

neg_all = sentiment_df[sentiment_df["rule_sentiment"] == "negative"]
remaining_budget = max(TARGET_N - len(neg_all), 0)
per_class_remaining = remaining_budget // 2

pos_available = sentiment_df[sentiment_df["rule_sentiment"] == "positive"]
neu_available = sentiment_df[sentiment_df["rule_sentiment"] == "neutral"]

pos_sample = pos_available.sample(n=min(per_class_remaining, len(pos_available)), random_state=42)
neu_sample = neu_available.sample(n=min(per_class_remaining, len(neu_available)), random_state=42)

annotation_sample = (
    pd.concat([neg_all, pos_sample, neu_sample])
    .sample(frac=1, random_state=42)   # shuffle so labelling order doesn't track rule_sentiment
    .reset_index(drop=True)
)
annotation_sample["gold_sentiment"] = ""

print("Annotation sample composition (target: roughly even across classes):")
print(annotation_sample["rule_sentiment"].value_counts())

export_cols = ["clean_text", "tourism_aspect", "sustainability_dimension", "rule_sentiment", "gold_sentiment"]
# include the model baseline as a cross-check column for the annotator, if Stage 2 has been run
if "model_sentiment" in annotation_sample.columns:
    export_cols.insert(-1, "model_sentiment")

annotation_sample[export_cols].to_csv("sentiment_annotation.csv", index=False)
print("\nSaved sentiment_annotation.csv - open it and fill 'gold_sentiment' with positive/neutral/negative for each row.")


Annotation sample composition (target: roughly even across classes):
rule_sentiment
negative    343
positive    328
neutral     328
Name: count, dtype: int64

Saved sentiment_annotation.csv - open it and fill 'gold_sentiment' with positive/neutral/negative for each row.


## Stage 4 — Fine-tune TourBERT (gated — do not "Run All" through this until labelling is done)

TourBERT (`veroman/TourBERT`) is a tourism-domain BERT **encoder** — it has no ready-made sentiment head, so it
must be fine-tuned on the gold labels from Stage 3 before it can be compared to anything. This is set behind an
explicit flag rather than a `try/except` around a missing file, so a top-to-bottom "Run All" never silently skips
training when labels *are* ready, and never crashes when they aren't:

1. Open `sentiment_annotation.csv`, fill in `gold_sentiment` for as many rows as you can (aim for 800–1,500;
   the stratified sample above already guarantees a reasonable number of each class).
2. Save the file, then flip `RUN_TOURBERT_TRAINING = True` below.

**Also requires internet + GPU (Colab), same as Stage 2.**

In [31]:
RUN_TOURBERT_TRAINING = False  # flip to True once sentiment_annotation.csv has gold_sentiment filled in

MIN_LABELLED_ROWS = 100  # below this, a fine-tuned model isn't meaningfully evaluable

if RUN_TOURBERT_TRAINING:
    labels = {"negative": 0, "neutral": 1, "positive": 2}

    labeled_df = pd.read_csv("sentiment_annotation.csv")
    labeled_df = labeled_df[labeled_df["gold_sentiment"].isin(labels.keys())].copy()

    if len(labeled_df) < MIN_LABELLED_ROWS:
        print(f"Only {len(labeled_df)} labelled rows found (need >= {MIN_LABELLED_ROWS}). "
              f"Label more of sentiment_annotation.csv before training - skipping for now.")
        RUN_TOURBERT_TRAINING = False
    else:
        labeled_df["label"] = labeled_df["gold_sentiment"].map(labels)
        print(labeled_df["gold_sentiment"].value_counts())
else:
    print("RUN_TOURBERT_TRAINING is False - skipping fine-tuning. "
          "This is expected until sentiment_annotation.csv has been hand-labelled.")


RUN_TOURBERT_TRAINING is False - skipping fine-tuning. This is expected until sentiment_annotation.csv has been hand-labelled.


In [32]:
if RUN_TOURBERT_TRAINING:
    from sklearn.model_selection import train_test_split
    from datasets import Dataset
    from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
    import numpy as np
    from sklearn.metrics import accuracy_score, precision_recall_fscore_support

    train_df, temp_df = train_test_split(
        labeled_df, test_size=0.30, stratify=labeled_df["label"], random_state=42
    )
    validation_df, test_df = train_test_split(
        temp_df, test_size=0.50, stratify=temp_df["label"], random_state=42
    )
    print(f"Train={len(train_df)}, Validation={len(validation_df)}, Test={len(test_df)}")

    TOURBERT = "veroman/TourBERT"
    tokenizer = AutoTokenizer.from_pretrained(TOURBERT)
    model = AutoModelForSequenceClassification.from_pretrained(
        TOURBERT, num_labels=3,
        id2label={0: "negative", 1: "neutral", 2: "positive"},
        label2id=labels,
    )
    # NOTE: a "classification layer newly initialized" warning here is expected -
    # TourBERT knows tourism language, the sentiment head is new and needs this fine-tuning.

    def tokenize(batch):
        return tokenizer(batch["clean_text"], truncation=True, padding="max_length", max_length=128)

    train_dataset = Dataset.from_pandas(train_df[["clean_text", "label"]]).map(tokenize, batched=True)
    validation_dataset = Dataset.from_pandas(validation_df[["clean_text", "label"]]).map(tokenize, batched=True)
    test_dataset = Dataset.from_pandas(test_df[["clean_text", "label"]]).map(tokenize, batched=True)

    def compute_metrics(pred):
        labels_true = pred.label_ids
        preds = np.argmax(pred.predictions, axis=1)
        precision, recall, f1, _ = precision_recall_fscore_support(
            labels_true, preds, average="macro", zero_division=0
        )
        return {"accuracy": accuracy_score(labels_true, preds),
                "precision_macro": precision, "recall_macro": recall, "f1_macro": f1}

    training_args = TrainingArguments(
        output_dir="./tourbert_sentiment",
        learning_rate=2e-5,
        per_device_train_batch_size=16,
        per_device_eval_batch_size=32,
        num_train_epochs=4,
        weight_decay=0.01,
        eval_strategy="epoch",
        save_strategy="epoch",
        load_best_model_at_end=True,
        metric_for_best_model="f1_macro",
        greater_is_better=True,
    )
    trainer = Trainer(
        model=model, args=training_args,
        train_dataset=train_dataset, eval_dataset=validation_dataset,
        tokenizer=tokenizer, compute_metrics=compute_metrics,
    )
    trainer.train()

    tourbert_test_results = trainer.evaluate(test_dataset)
    print(tourbert_test_results)

    # Apply the fine-tuned model to the full analysis-ready set (not just the labelled sample)
    full_dataset = Dataset.from_pandas(sentiment_df[["clean_text"]]).map(tokenize, batched=True)
    tourbert_logits = trainer.predict(full_dataset).predictions
    tourbert_probs = torch.softmax(torch.tensor(tourbert_logits), dim=1).numpy()
    id2label = {0: "negative", 1: "neutral", 2: "positive"}
    sentiment_df["tourbert_sentiment"] = [id2label[p.argmax()] for p in tourbert_probs]
    sentiment_df["tourbert_confidence"] = tourbert_probs.max(axis=1)
    sentiment_df.to_csv("sentiment_with_tourbert.csv", index=False)
    print("Saved sentiment_with_tourbert.csv")


## Stage 5 — Compare rule-based vs XLM-R vs fine-tuned TourBERT

**Macro F1, not accuracy, is the metric that matters here.** With a class split like 70% positive / 25% neutral /
5% negative, a model can post a deceptively high accuracy by mostly predicting "positive" — Macro F1 forces
reasonable performance on all three classes, including negative, which is exactly the class sustainability
problems show up in. Nothing is assumed in advance: if XLM-R matches or beats TourBERT on this data, use XLM-R.

In [33]:
if RUN_TOURBERT_TRAINING and "tourbert_test_results" in dir():
    from sklearn.metrics import accuracy_score, precision_recall_fscore_support

    def score_against_gold(gold_labels, pred_labels):
        precision, recall, f1, _ = precision_recall_fscore_support(
            gold_labels, pred_labels, average="macro", zero_division=0, labels=["negative","neutral","positive"]
        )
        _, neg_recall, _, _ = precision_recall_fscore_support(
            gold_labels, pred_labels, average=None, zero_division=0, labels=["negative","neutral","positive"]
        )[0:3]
        return {
            "accuracy": accuracy_score(gold_labels, pred_labels),
            "macro_f1": f1,
            "negative_recall": recall[0] if hasattr(recall, "__len__") else None,
        }

    gold_test = test_df["gold_sentiment"]
    comparison_rows = []
    for name, col in [("Keyword baseline", "rule_sentiment"), ("XLM-R", "model_sentiment")]:
        preds = sentiment_df.loc[test_df.index, col]
        comparison_rows.append({"model": name, **score_against_gold(gold_test, preds)})
    comparison_rows.append({
        "model": "Fine-tuned TourBERT",
        "accuracy": tourbert_test_results.get("eval_accuracy"),
        "macro_f1": tourbert_test_results.get("eval_f1_macro"),
        "negative_recall": None,  # per-class recall available via sklearn.metrics.classification_report if needed
    })
    comparison_table = pd.DataFrame(comparison_rows).round(3)
    print(comparison_table.to_string(index=False))
    print("\nDon't assume TourBERT wins - pick whichever row has the highest macro_f1 for FINAL_SENTIMENT_SOURCE below.")
else:
    print("Skipped - requires RUN_TOURBERT_TRAINING=True and a completed Stage 4 run "
          "(needs gold_sentiment labels in sentiment_annotation.csv).")


Skipped - requires RUN_TOURBERT_TRAINING=True and a completed Stage 4 run (needs gold_sentiment labels in sentiment_annotation.csv).


## Stage 6 — Choose the final sentiment source (auto-detects what's actually available)

Rather than hardcoding a choice, this picks the best available model in order of sophistication — fine-tuned
TourBERT if Stage 4/5 showed it's the strongest, else the XLM-R baseline, else the original rule-based labels — so
the same notebook works correctly whichever stages have actually been run.

In [34]:
if "tourbert_sentiment" in sentiment_df.columns:
    FINAL_SENTIMENT_SOURCE = "tourbert_sentiment"
    FINAL_CONFIDENCE_SOURCE = "tourbert_confidence"
elif "model_sentiment" in sentiment_df.columns:
    FINAL_SENTIMENT_SOURCE = "model_sentiment"
    FINAL_CONFIDENCE_SOURCE = "sentiment_confidence"
else:
    FINAL_SENTIMENT_SOURCE = "rule_sentiment"
    FINAL_CONFIDENCE_SOURCE = None

sentiment_df["final_sentiment"] = sentiment_df[FINAL_SENTIMENT_SOURCE]
sentiment_df["final_confidence"] = sentiment_df[FINAL_CONFIDENCE_SOURCE] if FINAL_CONFIDENCE_SOURCE else np.nan

print(f"Using '{FINAL_SENTIMENT_SOURCE}' as final_sentiment for all dashboard aggregation below.")
print(sentiment_df["final_sentiment"].value_counts())


Using 'model_sentiment' as final_sentiment for all dashboard aggregation below.
final_sentiment
positive    2115
negative    1612
neutral     1028
Name: count, dtype: int64


## Stage 7 — Dashboard-ready aggregation exports

Turns the 4,755 scored comments into the exact aggregated shapes a dashboard needs — never raw comments. Every
aggregation reports **both** a simple count-based `net_sentiment` and an **engagement-weighted** version (using the
`engagement_weight` = `log1p(like_count)` column that the original pipeline already computed but never used) so one
highly-liked comment doesn't get the same influence as one with zero likes, without letting it completely dominate
either.

$$\text{net\_sentiment} = \%positive - \%negative \quad (-100 = \text{all negative}, \; 0 = \text{balanced}, \; +100 = \text{all positive})$$

In [35]:
import os
os.makedirs("dashboard_exports", exist_ok=True)

SENT_COL = "final_sentiment"

def net_sentiment(s):
    c = s.value_counts(normalize=True) * 100
    return round(c.get("positive", 0.0) - c.get("negative", 0.0), 1)

def weighted_net_sentiment(labels, weights):
    pos_w = weights[labels == "positive"].sum()
    neg_w = weights[labels == "negative"].sum()
    tot_w = weights.sum()
    return round((pos_w - neg_w) / tot_w * 100, 1) if tot_w > 0 else np.nan


In [36]:
# ---- 7a. Overall headline KPI ----
overall_counts = sentiment_df[SENT_COL].value_counts()
overall_pct = (overall_counts / overall_counts.sum() * 100).round(1)

overall_summary = pd.DataFrame({
    "metric": ["analyzed_comments", "positive_pct", "neutral_pct", "negative_pct",
               "net_sentiment", "weighted_net_sentiment"],
    "value": [
        len(sentiment_df),
        overall_pct.get("positive", 0.0),
        overall_pct.get("neutral", 0.0),
        overall_pct.get("negative", 0.0),
        net_sentiment(sentiment_df[SENT_COL]),
        weighted_net_sentiment(sentiment_df[SENT_COL], sentiment_df["engagement_weight"]),
    ]
})
print(overall_summary.to_string(index=False))
overall_summary.to_csv("dashboard_exports/agg_overall.csv", index=False)


                metric  value
     analyzed_comments 4755.0
          positive_pct   44.5
           neutral_pct   21.6
          negative_pct   33.9
         net_sentiment   10.6
weighted_net_sentiment   19.5


In [37]:
# ---- 7b. Sustainability dimension ----
dim_df = sentiment_df.assign(
    sustainability_dimension=sentiment_df["sustainability_dimension"].str.split(";")
).explode("sustainability_dimension")
dim_df = dim_df[dim_df["sustainability_dimension"] != "none"]

dim_agg = dim_df.groupby("sustainability_dimension").apply(
    lambda g: pd.Series({
        "n": len(g),
        "positive_pct": round((g[SENT_COL] == "positive").mean() * 100, 1),
        "neutral_pct": round((g[SENT_COL] == "neutral").mean() * 100, 1),
        "negative_pct": round((g[SENT_COL] == "negative").mean() * 100, 1),
        "net_sentiment": net_sentiment(g[SENT_COL]),
        "weighted_net_sentiment": weighted_net_sentiment(g[SENT_COL], g["engagement_weight"]),
    })
).reset_index().sort_values("net_sentiment")

print(dim_agg.to_string(index=False))
dim_agg.to_csv("dashboard_exports/agg_dimension.csv", index=False)


sustainability_dimension      n  positive_pct  neutral_pct  negative_pct  net_sentiment  weighted_net_sentiment
          infrastructure 1978.0          31.3         26.3          42.4          -11.1                    -1.9
                  social 1134.0          48.2         12.7          39.1            9.2                    20.3
           environmental 1454.0          43.2         24.0          32.8           10.4                    13.3
                economic 2268.0          50.2         18.6          31.3           18.9                    24.0


/tmp/ipykernel_1074/3435654878.py:7: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  dim_agg = dim_df.groupby("sustainability_dimension").apply(


In [38]:
# ---- 7c. Tourism aspect (powers a "Top Negative Issues" ranking) ----
MIN_N_FOR_RANKING = 20  # hide aspects with too few comments to trust a percentage

asp_df = sentiment_df.assign(
    tourism_aspect=sentiment_df["tourism_aspect"].str.split(";")
).explode("tourism_aspect")
asp_df = asp_df[asp_df["tourism_aspect"] != "none"]

asp_agg = asp_df.groupby("tourism_aspect").apply(
    lambda g: pd.Series({
        "n": len(g),
        "positive_pct": round((g[SENT_COL] == "positive").mean() * 100, 1),
        "negative_pct": round((g[SENT_COL] == "negative").mean() * 100, 1),
        "net_sentiment": net_sentiment(g[SENT_COL]),
        "weighted_net_sentiment": weighted_net_sentiment(g[SENT_COL], g["engagement_weight"]),
    })
).reset_index()
asp_agg["low_sample_warning"] = asp_agg["n"] < MIN_N_FOR_RANKING

print("Top negative-leaning aspects (min n=%d):" % MIN_N_FOR_RANKING)
print(asp_agg[~asp_agg["low_sample_warning"]].sort_values("negative_pct", ascending=False).head(8).to_string(index=False))
asp_agg.to_csv("dashboard_exports/agg_aspect.csv", index=False)


Top negative-leaning aspects (min n=20):
tourism_aspect      n  positive_pct  negative_pct  net_sentiment  weighted_net_sentiment  low_sample_warning
   cleanliness  199.0          33.7          61.8          -28.1                    -7.3               False
    facilities  136.0          22.1          61.0          -39.0                   -40.8               False
        safety  241.0          36.9          54.4          -17.4                    -9.3               False
 accessibility   74.0          35.1          52.7          -17.6                   -22.3               False
     transport 1305.0          25.0          51.6          -26.6                   -14.7               False
   hospitality  411.0          43.6          51.6           -8.0                    12.2               False
   price_value  495.0          36.8          44.6           -7.9                     1.4               False
      crowding   91.0          40.7          42.9           -2.2                     2.

/tmp/ipykernel_1074/1483371773.py:9: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  asp_agg = asp_df.groupby("tourism_aspect").apply(


In [39]:
# ---- 7d. Destination x dimension heatmap ----
MIN_N_FOR_DESTINATION = 20

dd_df = sentiment_df.assign(
    destination=sentiment_df["destination"].str.split(";"),
    sustainability_dimension=sentiment_df["sustainability_dimension"].str.split(";"),
).explode("destination").explode("sustainability_dimension")
dd_df = dd_df[(dd_df["destination"] != "unknown") & (dd_df["sustainability_dimension"] != "none")]

dd_agg = dd_df.groupby(["destination", "sustainability_dimension"]).apply(
    lambda g: pd.Series({"n": len(g), "net_sentiment": net_sentiment(g[SENT_COL])})
).reset_index()

dest_totals = dd_df.groupby("destination").size()
keep_destinations = dest_totals[dest_totals >= MIN_N_FOR_DESTINATION].index
dd_agg_filtered = dd_agg[dd_agg["destination"].isin(keep_destinations)]

heatmap = dd_agg_filtered.pivot(index="destination", columns="sustainability_dimension", values="net_sentiment")
heatmap_n = dd_agg_filtered.pivot(index="destination", columns="sustainability_dimension", values="n")

print(f"Destination x dimension net-sentiment heatmap ({len(keep_destinations)} destinations with n>={MIN_N_FOR_DESTINATION}):")
print(heatmap.round(1))
dd_agg.to_csv("dashboard_exports/agg_destination_dimension.csv", index=False)


Destination x dimension net-sentiment heatmap (12 destinations with n>=20):
sustainability_dimension  economic  environmental  infrastructure  social
destination                                                              
Johor                         26.8           39.4            27.7    33.3
Kuala Lumpur                  20.8           25.3             5.0    12.2
Langkawi                      25.0           45.2            47.4    65.2
Malaysia                       6.4          -24.2           -40.6   -13.6
Melaka                        36.6            9.3            23.6    37.8
Pahang                        21.1           19.4            25.0    16.7
Penang                        38.1           35.8            37.1    45.1
Perak                         66.7           69.2            64.3    88.9
Putrajaya                     16.7            0.0           -25.0    10.0
Sabah                         63.2           49.2            40.7    47.2
Sarawak                       45.3  

/tmp/ipykernel_1074/690803081.py:10: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  dd_agg = dd_df.groupby(["destination", "sustainability_dimension"]).apply(


In [40]:
# ---- 7e. Stakeholder (local / visitor / potential_visitor / unknown) ----
LOW_SAMPLE_THRESHOLD = 30

stake_agg = sentiment_df.groupby("stakeholder").apply(
    lambda g: pd.Series({
        "n": len(g),
        "positive_pct": round((g[SENT_COL] == "positive").mean() * 100, 1),
        "neutral_pct": round((g[SENT_COL] == "neutral").mean() * 100, 1),
        "negative_pct": round((g[SENT_COL] == "negative").mean() * 100, 1),
        "net_sentiment": net_sentiment(g[SENT_COL]),
    })
).reset_index()
# Machine-readable flag (not just a note a developer has to remember) so the dashboard
# can auto-render an "n=X, interpret with caution" caveat next to small groups.
stake_agg["low_sample_warning"] = stake_agg["n"] < LOW_SAMPLE_THRESHOLD

print(stake_agg.to_string(index=False))
stake_agg.to_csv("dashboard_exports/agg_stakeholder.csv", index=False)


      stakeholder      n  positive_pct  neutral_pct  negative_pct  net_sentiment  low_sample_warning
            local  168.0          35.1          6.0          58.9          -23.8               False
potential_visitor   28.0          82.1         17.9           0.0           82.1                True
          unknown 4541.0          44.5         22.3          33.3           11.2               False
          visitor   18.0          77.8         11.1          11.1           66.7                True


/tmp/ipykernel_1074/3224530391.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  stake_agg = sentiment_df.groupby("stakeholder").apply(


In [41]:
# ---- 7f. Sentiment over time ----
# NOTE: 2026 dates are NOT excluded here (see the Data Quality Note at the top) - if you
# decide to apply that filter upstream, this will automatically reflect it on next run.
ts_df = sentiment_df.copy()
ts_df["comment_published_at"] = pd.to_datetime(ts_df["comment_published_at"], errors="coerce", utc=True).dt.tz_localize(None)
ts_df["year_month"] = ts_df["comment_published_at"].dt.to_period("M").astype(str)
ts_df = ts_df.assign(
    sustainability_dimension=ts_df["sustainability_dimension"].str.split(";")
).explode("sustainability_dimension")
ts_df = ts_df[ts_df["sustainability_dimension"] != "none"]

ts_agg = ts_df.groupby(["year_month", "sustainability_dimension"]).apply(
    lambda g: pd.Series({"n": len(g), "net_sentiment": net_sentiment(g[SENT_COL])})
).reset_index()
ts_agg["low_sample_warning"] = ts_agg["n"] < 10

sparse_share = ts_agg["low_sample_warning"].mean()
print(f"{sparse_share:.0%} of year_month x dimension buckets have n<10 - monthly granularity is sparse. "
      f"Consider aggregating to quarters in the dashboard for a more stable trend line; "
      f"n and low_sample_warning are exported so the frontend can decide.")
ts_agg.to_csv("dashboard_exports/agg_timeseries_monthly.csv", index=False)

# Quarterly rollup, exported alongside monthly for a steadier default view
ts_df["year_quarter"] = ts_df["comment_published_at"].dt.to_period("Q").astype(str)
tsq_agg = ts_df.groupby(["year_quarter", "sustainability_dimension"]).apply(
    lambda g: pd.Series({"n": len(g), "net_sentiment": net_sentiment(g[SENT_COL])})
).reset_index()
tsq_agg.to_csv("dashboard_exports/agg_timeseries_quarterly.csv", index=False)
print(tsq_agg.tail(8).to_string(index=False))


49% of year_month x dimension buckets have n<10 - monthly granularity is sparse. Consider aggregating to quarters in the dashboard for a more stable trend line; n and low_sample_warning are exported so the frontend can decide.
year_quarter sustainability_dimension      n  net_sentiment
      2026Q3                 economic  166.0           35.5
      2026Q3            environmental   84.0            9.5
      2026Q3           infrastructure  106.0            4.7
      2026Q3                   social   80.0           12.5
         NaT                 economic 1080.0            9.1
         NaT            environmental  544.0           -6.8
         NaT           infrastructure 1037.0          -35.4
         NaT                   social  534.0           -7.3


/tmp/ipykernel_1074/1809170993.py:12: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  ts_agg = ts_df.groupby(["year_month", "sustainability_dimension"]).apply(
/tmp/ipykernel_1074/1809170993.py:25: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tsq_agg = ts_df.groupby(["year_quarter", "sustainability_dimension"]).apply(


In [42]:
# ---- Also export the full scored comment-level dataset (for drill-down / search in the dashboard) ----
drilldown_cols = [
    "comment_id", "video_id", "clean_text", "comment_published_at", "like_count",
    "destination", "stakeholder", "tourism_aspect", "sustainability_dimension",
    "rule_sentiment", SENT_COL, "final_confidence", "helpfulness_score", "engagement_weight",
]
drilldown_cols = [c for c in drilldown_cols if c in sentiment_df.columns]
sentiment_df[drilldown_cols].to_csv("dashboard_exports/comments_scored.csv", index=False)

print("All dashboard exports written to ./dashboard_exports/:")
for f in sorted(os.listdir("dashboard_exports")):
    print(" -", f)


All dashboard exports written to ./dashboard_exports/:
 - agg_aspect.csv
 - agg_destination_dimension.csv
 - agg_dimension.csv
 - agg_overall.csv
 - agg_stakeholder.csv
 - agg_timeseries_monthly.csv
 - agg_timeseries_quarterly.csv
 - comments_scored.csv


## Stage 8 — Roadmap: Aspect-Based Sentiment Analysis (not implemented here)

*"The food was amazing but traffic in KL was horrible"* currently gets **one** `final_sentiment` label for the
whole comment, and the aggregations above give it credit in both the `food` and `transport` aspect buckets under
that same single label — a comment-level approximation, not true phrase-level ABSA. This is workable for a first
version (it's why `tourism_aspect` is captured per comment already), but the next real upgrade is a proper ABSA
model that assigns a *separate* sentiment to each aspect mentioned in the same comment
(`food → positive, transport → negative`), which would directly sharpen the destination heatmap and the
top-negative-issues ranking above. Scoping and implementing that is a separate follow-up, not part of this
refinement.

## Summary: what's ready to hand to the dashboard

| Export | Feeds |
|---|---|
| `agg_overall.csv` | `Overview` — headline sentiment KPI cards |
| `agg_dimension.csv` | `Overview`/`States` — sustainability-dimension bars |
| `agg_aspect.csv` | `States` — "top negative issues" ranking |
| `agg_destination_dimension.csv` | `States` — per-state pillar-sentiment detail, alongside the STI pillar chart |
| `agg_stakeholder.csv` | `Overview`/`Methodology` — resident vs visitor split, with built-in low-sample flags |
| `agg_timeseries_monthly.csv` / `_quarterly.csv` | `Overview` — sentiment trend line, alongside the CPI/visitor trend already there |
| `comments_scored.csv` | drill-down / search behind any chart |

Every export carries an `n` (and, where relevant, `low_sample_warning`) column so the frontend can render a
sample-size caveat automatically instead of a developer having to remember to add one by hand.

##download

In [50]:
import base64
from IPython.display import HTML, display

zip_filename = 'dashboard_exports.zip'

# Read the zip file and encode it in base64
with open(zip_filename, 'rb') as f:
    b64 = base64.b64encode(f.read()).decode('utf-8')

# Create a client-side HTML download button
html_code = f'''
<a download="{zip_filename}" href="data:application/zip;base64,{b64}" style="padding: 10px 20px; background-color: #007bff; color: white; text-decoration: none; border-radius: 5px; font-weight: bold; display: inline-block;">
    Click here to download {zip_filename}
</a>
'''
display(HTML(html_code))